redaxo/src/addons/structure/lib/navigation_iterator.php

```diff
@@ -0,0 +1,298 @@
+<?php
+
+/**
+ * Klasse zum Erstellen von Navigationen, v0.1.
+ *
+ * benötigt PHP7!
+ *
+ * @package redaxo\structure
+ */
+
+/*
+ * Beispiel-Konfiguration einer Haupt-Navigation:
+ *
+ * $it = rex_navigation_iterator::factory()
+ *   ->ignoreOfflines(true)
+ * ;
+ *
+ * foreach($it as $navItem) {
+ *    // @var rex_article|rex_category $ooObj das aktuelle Element 
+ *    // @var int $depth die aktuelle tiefe/ebene 
+ *    // @var int $nth ein forlaufender Zähler je Ebene 
+ *    // @var int $count Anzahl Elemente auf der aktuellen Ebene 
+ *    list($ooObj, $depth, $nth, $count) = $navItem;
+ *
+ *    if ($nth == 1) {
+ *        if ($depth == 1) {
+ *           $class = "nav navbar-nav";
+ *        } else {
+ *           $class = "dropdown";
+ *        }
+ *        echo '<ul class="'. $class .'">';
+ *    }
+ *
+ *    echo "<li class='rex-navi-depth-". $depth ." rex-nav-". $nth ."nth-child'>";
+ *    echo "<a href='". $ooObj->getUrl() ."'>". $ooObj->getName() ."</a>";
+ *    echo "</li>\n";
```

Comment:
@staabm Ich sehe hier konzeptionell noch ein Grundproblem (siehe auch Diskussion unter dem Gist).
Die Unterliste (`<ul>`) muss mit in das `<li>`, sonst erhält man keinen validen Code. 
Ich sehe aber aktuell keine Möglichkeit, wie man das mit der Klasse erreichen kann, da die Unterliste erst "abgearbeitet" wird, wenn der Oberpunkt fertig durchlaufen ist. Somit bekommt man die Unterliste nicht vor dieses `</li>` (falls ich nicht was Wesentliches übersehe).

Reply:
Genau. Sie muss ins `<li>`, darf aber nicht in den Anker. Und: Es muss auch die Möglichkeit geben, Navigationspunkte ohne Anker zu generieren, bspw. nur mit einem `<span>` umschlossen.

Reply:
Also das markup generiert ja jeder selbst der die klasse benutzt. Es ist nicht so wie bei der navigation-factory wo die klasse das markup vorgibt. D.h. Jeder generiet das markup was er brauch. Das hier sind nur beispiele, man muss die klasse nicht mit dem hier dargestellten markup verwenden.

Reply:
Tja, dann, sind wir wieder am Anfang: praxisnahe und gültige Beispiele...

Reply:
@staabm Ich habe den Eindruck, dass du die Problematik noch nicht ganz siehst, die @tbaddade unter dem gist aufgezeigt hat.
Ich sehe aktuell keine Möglichkeit, mit der Klasse eine sinnvolle/valide ul-li-Struktur über 2 oder mehr Ebenen zu erstellen. 
Falls das doch möglich ist, am besten mal beispielhaft skizzieren (und am besten das obige Beispiel damit austauschen, denn da kommt nichts sinnvolles raus, sobald man mehrere Ebenen hat).

Reply:
Ich kann das Problem auch bestätigen. Es ist mir nicht aufgefallen, da ich für die Lösungen wo ich es benutze sehr flache Navigationen verwende. Zusätzlich zur depth, benötigt man evtl. noch eine Info haschilds oder Ähnliches damit man direkt reagieren kann. Andernfalls geht's nur, wenn man sich anhand der Daten der gelieferten Structure Elemente selbst noch mal ein Array baut.